samza-sql/src/main/java/org/apache/samza/sql/data/RexToJavaCompiler.java

```diff
@@ -0,0 +1,241 @@
+ */
+public class RexToJavaCompiler {
+  private static final Logger log = LoggerFactory.getLogger(RexToJavaCompiler.class);
+
+  private final RexBuilder rexBuilder;
+
+  public RexToJavaCompiler(RexBuilder rexBuilder) {
+    this.rexBuilder = rexBuilder;
+  }
+
+  /**
+   * Compiles a row expression to a instance of {@link org.apache.samza.sql.data.Expression}
+   * @param inputs Input relations/time-varying relations for this row expression
+   * @param nodes row expression
+   * @return compiled expression of type {@link org.apache.samza.sql.data.Expression}
+   */
+  public org.apache.samza.sql.data.Expression compile(List<RelNode> inputs, List<RexNode> nodes) {
+    /*
+       In case there are multiple input relations, we build a single input row type combining types of all the inputs.
+     */
+    final RelDataTypeFactory.FieldInfoBuilder fieldBuilder = rexBuilder.getTypeFactory().builder();
+    for (RelNode input : inputs) {
+      fieldBuilder.addAll(input.getRowType().getFieldList());
+    }
+    final RelDataType inputRowType = fieldBuilder.build();
+    final RexProgramBuilder programBuilder = new RexProgramBuilder(inputRowType, rexBuilder);
+    for (RexNode node : nodes) {
+      programBuilder.addProject(node, null);
+    }
+    final RexProgram program = programBuilder.getProgram();
+
+    final BlockBuilder builder = new BlockBuilder();
+    final ParameterExpression inputValues = Expressions.parameter(Object[].class, "inputValues");
+    final ParameterExpression outputValues = Expressions.parameter(Object[].class, "outputValues");
+    final JavaTypeFactoryImpl javaTypeFactory = new JavaTypeFactoryImpl(rexBuilder.getTypeFactory().getTypeSystem());
+
+    // public void execute(Object[] inputValues, Object[] outputValues)
+    final RexToLixTranslator.InputGetter inputGetter = new RexToLixTranslator.InputGetterImpl(ImmutableList.of(
+        Pair.<org.apache.calcite.linq4j.tree.Expression, PhysType>of(
+            Expressions.variable(Object[].class, "inputValues"),
+            PhysTypeImpl.of(javaTypeFactory, inputRowType, JavaRowFormat.ARRAY, false))));
+    final Function1<String, RexToLixTranslator.InputGetter> correlates = a0 -> {
+      throw new UnsupportedOperationException();
+    };
+    final List<org.apache.calcite.linq4j.tree.Expression> list =
+        RexToLixTranslator.translateProjects(program, javaTypeFactory, builder, null, null, inputGetter, correlates);
+    for (int i = 0; i < list.size(); i++) {
+      builder.add(Expressions.statement(
+          Expressions.assign(Expressions.arrayIndex(outputValues, Expressions.constant(i)), list.get(i))));
+    }
+    return createSamzaExpressionFromCalcite(inputValues, outputValues, builder.toBlock());
+  }
+
+  /**
+   * Given a method that implements {@link org.apache.samza.sql.data.Expression#execute(Object[], Object[])},
+   * adds a bridge method that implements {@link org.apache.samza.sql.data.Expression#execute(Object[])}, and
```

Comment:
What is this bridge method used for? Just make sure we have a dummy implementation of Object execute(Object[] inputValues)? Or actually binding fields in the input row to Java objects?